workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs

```diff
@@ -1,53 +1,127 @@
+        private static readonly string InstallMarkerFile =
+            Path.GetFullPath($"build/CoreSdk/{Common.CoreSdkVersion}.installed");
+
+        private static readonly string ProjectPath =
+            Path.GetFullPath(Path.Combine(Common.GetThisPackagePath(), ".DownloadCoreSdk/DownloadCoreSdk.csproj"));
+
+        private const string DownloadForceMenuItem = "Improbable/Download CoreSdk (force)";
+
+        [MenuItem(DownloadForceMenuItem, false, DownloadForcePriority)]
+        private static void DownloadForceMenu()
         {
             Download();
             AssetDatabase.Refresh();
         }
 
-        [MenuItem("Improbable/Download CoreSdk (force)", false, DownloadForcePriority)]
-        public static void DownloadForceMenu()
+        /// <summary>
+        ///     Windows only: Ensures that the user can't try to force a download if any of Improbable's native plugins are loaded
+        ///     by Unity.
+        /// </summary>
+        [MenuItem(DownloadForceMenuItem, true, DownloadForcePriority)]
+        private static bool DownloadForceValidate()
         {
-            var path = Path.Combine("build", "CoreSdk");
-            if (Directory.Exists(path))
+            var failedPlugins = new List<PluginImporter>();
+
+            // Unity never unloads native plugins. Detect them here and give a more useful error message.
+            foreach (var plugin in PluginImporter.GetAllImporters().Where(ShouldCheckPluginForLock))
             {
-                Directory.Delete(path, true);
+                var path = plugin.assetPath;
+                try
+                {
+                    // Try to open the file to see if it's locked.
+                    using (new FileStream(path, FileMode.Open, FileAccess.Write, FileShare.None))
+                    {
+                    }
+                }
+                catch
+                {
+                    failedPlugins.Add(plugin);
+                }
             }
 
-            DownloadMenu();
+            return !failedPlugins.Any();
         }
 
-        public static void Download()
+        private static bool ShouldCheckPluginForLock(PluginImporter p)
+        {
+            if (!p.assetPath.StartsWith("Assets/Plugins/Improbable"))
+            {
+                return false;
+            }
+
+            return p.isNativePlugin && File.Exists(p.assetPath);
+        }
+
+        /// <summary>
+        ///     Downloads the Core Sdk only if it has not already been downloaded and unzipped.
+        /// </summary>
+        public static DownloadResult TryDownload()
+        {
+            if (File.Exists(InstallMarkerFile))
+            {
+                return DownloadResult.AlreadyInstalled;
+            }
+
+            return Download();
+        }
+
+        /// <summary>
+        ///     Downloads and extracts the Core Sdk and associated libraries and tools.
+        /// </summary>
+        public static DownloadResult Download()
```

Comment:
Does this and the function above it need to be public?

Reply:
Not any more!